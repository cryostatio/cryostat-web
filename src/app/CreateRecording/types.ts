/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { EventTemplate, KeyValue } from '@app/Shared/Services/api.types';
import { ValidatedOptions } from '@patternfly/react-core';

export type EventTemplateIdentifier = Pick<EventTemplate, 'name' | 'type'>;

interface _FormBaseData {
  name: string;
  template?: EventTemplateIdentifier;
  labels: KeyValue[];
  continuous: boolean;
  archiveOnStop: boolean;
  restart: boolean;
  duration: number;
  durationUnit: number;
  skipDurationCheck?: boolean;
  maxAge: number;
  maxAgeUnit: number;
  maxSize: number;
  maxSizeUnit: number;
  toDisk: boolean;
}

interface _FormValidationData {
  nameValid: ValidatedOptions;
  labelsValid: ValidatedOptions;
  durationValid: ValidatedOptions;
}

export type CustomRecordingFormData = _FormBaseData & _FormValidationData;
