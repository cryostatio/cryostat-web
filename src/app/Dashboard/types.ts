/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { FeatureLevel } from '@app/Shared/Services/service.types';
import { gridSpans, LabelProps } from '@patternfly/react-core';
import { Observable } from 'rxjs';

export interface Sized<T> {
  minimum: T;
  default: T;
  maximum: T;
}

export interface DashboardCardSizes {
  span: Sized<gridSpans>;
  height: Sized<number>;
}

export interface DashboardCardDescriptor {
  featureLevel: FeatureLevel;
  icon?: React.ReactNode;
  labels?: {
    content: string;
    color?: LabelProps['color'];
    icon?: React.ReactNode;
  }[];
  preview?: React.ReactNode;
  title: string;
  cardSizes: DashboardCardSizes;
  description: string;
  descriptionFull: JSX.Element | string;
  component: DashboardCardFC<DashboardCardTypeProps>;
  propControls: PropControl[];
  advancedConfig?: JSX.Element;
}
export type DashboardCardFC<P> = React.FC<P> & {
  cardComponentName: string;
};

export interface PropControlExtra {
  displayMapper?: (value: string) => string /* only has effect with 'select' PropControl kind */;
  min?: number;
  max?: number;
  [key: string]: unknown;
}

export interface PropControl {
  name: string;
  key: string;
  description: string;
  kind: 'boolean' | 'number' | 'string' | 'text' | 'select';
  values?: unknown[] | Observable<unknown>;
  defaultValue: unknown;
  extras?: PropControlExtra;
}

export interface DashboardCardTypeProps {
  span: number;
  dashboardId: number;
  isDraggable?: boolean;
  isResizable?: boolean;
  isFullHeight?: boolean;
  actions?: JSX.Element[];
}

export interface ValidationError {
  message: React.ReactNode;
}

export interface CardValidationResult {
  errors: ValidationError[];
  callForAction?: React.ReactNode;
}

export type LayoutTemplateFilter = 'Suggested' | 'Cryostat' | 'User-submitted';

export interface LayoutTemplateController {
  selectedTemplate: SelectedLayoutTemplate | undefined;
  setSelectedTemplate: (template: React.SetStateAction<SelectedLayoutTemplate | undefined>) => void;
  isUploadModalOpen: boolean;
  setIsUploadModalOpen: (isOpen: React.SetStateAction<boolean>) => void;
}

export interface CardConfig {
  id: string;
  name: string;
  span: gridSpans;
  props: object;
}

export type SerialCardConfig = Omit<CardConfig, 'id'>;

export const mockSerialCardConfig: SerialCardConfig = {
  name: 'Default',
  span: 12,
  props: {},
};
export interface DashboardLayout {
  name: string;
  cards: CardConfig[];
  favorite: boolean;
}

export type SerialDashboardLayout = Omit<DashboardLayout, 'name' | 'cards'> & { cards: SerialCardConfig[] };

// only name and vendor are needed to identify a template
export type LayoutTemplateRecord = Pick<LayoutTemplate, 'name' | 'vendor'>;

export enum LayoutTemplateVersion {
  'v2.3' = 'v2.3',
  'v2.4' = 'v2.4',
}

export enum LayoutTemplateVendor {
  BLANK = 'Blank',
  CRYOSTAT = 'Cryostat',
  USER = 'User-submitted',
}

export interface LayoutTemplate {
  name: string;
  description: string;
  cards: SerialCardConfig[];
  version: LayoutTemplateVersion;
  vendor: LayoutTemplateVendor;
}

export interface SelectedLayoutTemplate {
  template: LayoutTemplate;
  category: LayoutTemplateFilter;
}

export type SerialLayoutTemplate = Omit<LayoutTemplate, 'vendor'>;
