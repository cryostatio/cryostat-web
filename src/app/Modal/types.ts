/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export enum DeleteOrDisableWarningType {
  DeleteActiveRecordings = 'DeleteActiveRecordings',
  DeleteArchivedRecordings = 'DeleteArchivedRecordings',
  DeleteAutomatedRules = 'DeleteAutomatedRules',
  DisableAutomatedRules = 'DisableAutomatedRules',
  DeleteEventTemplates = 'DeleteEventTemplates',
  DeleteProbeTemplates = 'DeleteProbeTemplates',
  DeleteActiveProbes = 'DeleteActiveProbes',
  DeleteCredentials = 'DeleteCredentials',
  DeleteCustomTargets = 'DeleteCustomTargets',
  DeleteDashboardLayout = 'DeleteDashboardLayout',
  DeleteLayoutTemplate = 'DeleteLayoutTemplate',
  ClearDashboardLayout = 'ClearDashboardLayout',
}

export interface DeleteOrDisableWarning {
  id: DeleteOrDisableWarningType;
  title: string;
  label: string;
  description: string;
  ariaLabel: string;
}

export const DeleteActiveRecordings: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.DeleteActiveRecordings,
  title: 'Permanently delete your JFR recording?',
  label: 'Delete Active Recording',
  description: `If you click delete, recording and report data will be lost.`,
  ariaLabel: 'Recording delete warning',
};

export const DeleteArchivedRecordings: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.DeleteArchivedRecordings,
  title: 'Permanently delete your archived JFR recording?',
  label: 'Delete Archived Recording',
  description: `If you click delete, recording and report data will be lost.`,
  ariaLabel: 'Recording delete warning',
};

export const DeleteAutomatedRules: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.DeleteAutomatedRules,
  title: 'Permanently delete your Automated Rule?',
  label: 'Delete Automated Rule',
  description: `If you click delete, rule data will be lost.`,
  ariaLabel: 'Automated rule delete warning',
};

export const DisableAutomatedRules: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.DisableAutomatedRules,
  title: 'Disable your Automated Rule?',
  label: 'Disable Automated Rule',
  description: `If you click Disable, the rule will be disabled.`,
  ariaLabel: 'Automated rule disable warning',
};

export const DeleteEventTemplates: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.DeleteEventTemplates,
  title: 'Permanently delete your Event Template?',
  label: 'Delete Event Template',
  description: `If you click delete, custom event template data will be lost.`,
  ariaLabel: 'Event template delete warning',
};

export const DeleteProbeTemplates: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.DeleteProbeTemplates,
  title: 'Permanently delete your Probe Template?',
  label: 'Delete Probe Template',
  description: `If you click delete, custom probe template data will be lost.`,
  ariaLabel: 'Probe template delete warning',
};

export const DeleteActiveProbes: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.DeleteActiveProbes,
  title: 'Permanently remove your Active Probes from the target?',
  label: 'Remove Active Probes',
  description: `If you click delete, active probes will be removed from the target.`,
  ariaLabel: 'Active Probes remove warning',
};

export const DeleteCredentials: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.DeleteCredentials,
  title: 'Permanently delete your Credentials?',
  label: 'Delete Credentials',
  description: `If you click delete, credential data for this target will be lost.`,
  ariaLabel: 'Credentials delete warning',
};

export const DeleteCustomTargets: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.DeleteCustomTargets,
  title: 'Permanently delete your Custom Target?',
  label: 'Delete Custom Targets',
  description: `If you click delete, custom target information will be lost.`,
  ariaLabel: 'Custom Targets delete warning',
};

export const DeleteDashboardLayout: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.DeleteDashboardLayout,
  title: 'Permanently delete your Dashboard Layout?',
  label: 'Delete Dashboard Layouts',
  description: `If you click delete, dashboard layout configuration data will be lost.`,
  ariaLabel: 'Dashboard Layout delete warning',
};

export const ClearDashboardLayout: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.ClearDashboardLayout,
  title: 'Permanently clear your Dashboard Layout?',
  label: 'Clear Dashboard Layout',
  description: 'If you click clear, all cards in the current dashboard layout will be removed.',
  ariaLabel: 'Dashboard Layout clear warning',
};

export const DeleteLayoutTemplate: DeleteOrDisableWarning = {
  id: DeleteOrDisableWarningType.DeleteLayoutTemplate,
  title: 'Permanently delete your Layout Template?',
  label: 'Delete Layout Template',
  description: `If you click delete, layout template configuration data will be lost.`,
  ariaLabel: 'Layout Template delete warning',
};

export const DeleteWarningKinds: DeleteOrDisableWarning[] = [
  DeleteActiveRecordings,
  DeleteArchivedRecordings,
  DeleteAutomatedRules,
  DisableAutomatedRules,
  DeleteEventTemplates,
  DeleteProbeTemplates,
  DeleteActiveProbes,
  DeleteCredentials,
  DeleteCustomTargets,
  DeleteDashboardLayout,
  DeleteLayoutTemplate,
  ClearDashboardLayout,
];
