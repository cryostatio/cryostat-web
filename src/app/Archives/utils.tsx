/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RecordingDirectory, Target } from '@app/Shared/Services/api.types';

export const includesDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): boolean => {
  return arr.some((t) => t.connectUrl === dir.connectUrl);
};

export const indexOfDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): number => {
  let index = -1;
  arr.forEach((d, idx) => {
    if (d.connectUrl === dir.connectUrl) {
      index = idx;
    }
  });
  return index;
};

export const getTargetFromDirectory = (dir: RecordingDirectory): Target => {
  return {
    connectUrl: dir.connectUrl,
    alias: dir.jvmId,
    labels: [],
    annotations: {
      cryostat: [],
      platform: [],
    },
  };
};
