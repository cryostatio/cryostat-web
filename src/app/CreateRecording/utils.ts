/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { TemplateType } from '@app/Shared/Services/api.types';
import { EventTemplateIdentifier } from './types';

export const RecordingNamePattern = /^[\w_]+$/;

export const DurationPattern = /^[1-9][0-9]*$/;

export const isRecordingNameValid = (name: string) => RecordingNamePattern.test(name);

export const isDurationValid = (duration: number) => DurationPattern.test(`${duration}`);

export const templateFromEventSpecifier = (eventSpecifier: string): EventTemplateIdentifier | undefined => {
  const regex = /^template=([a-zA-Z0-9]+)(?:,type=([a-zA-Z0-9]+))?$/im;
  if (eventSpecifier && regex.test(eventSpecifier)) {
    const parts = regex.exec(eventSpecifier);
    if (parts) {
      return {
        name: parts[1],
        type: parts[2] as TemplateType,
      };
    }
  }
  return undefined;
};
