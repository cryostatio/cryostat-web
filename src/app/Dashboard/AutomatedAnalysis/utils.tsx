/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { AnalysisResult, Evaluation } from '@app/Shared/Services/api.types';
import { Stack, StackItem, Label, Title, Text } from '@patternfly/react-core';
import _ from 'lodash';

export const transformAADescription = (result: AnalysisResult): JSX.Element => {
  const format = (s: Evaluation): JSX.Element => {
    if (typeof s === 'string') {
      return <Text>{s}</Text>;
    }
    if (Array.isArray(s)) {
      return (
        <Stack>
          {s.map((e) => (
            <StackItem key={e.setting}>
              <Title headingLevel={'h6'}>{e.setting}</Title>
              <Label>
                {e.name}={e.value}
              </Label>
            </StackItem>
          ))}
        </Stack>
      );
    }
    throw `Unrecognized item: ${s}`;
  };
  return (
    <div>
      {Object.entries(result.evaluation || {}).map(([k, v]) =>
        v && v.length ? (
          <div key={k}>
            <span>
              <Title headingLevel={'h5'}>{_.capitalize(k)}</Title>
              {format(result.evaluation[k])}
            </span>
            <br />
          </div>
        ) : (
          <div key={k}></div>
        ),
      )}
    </div>
  );
};
