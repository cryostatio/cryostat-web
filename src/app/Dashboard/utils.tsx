/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import cryostatLogo from '@app/assets/cryostat_icon_rgb_default.svg';
import cryostatLogoDark from '@app/assets/cryostat_icon_rgb_reverse.svg';
import { dashboardConfigDeleteCardIntent } from '@app/Shared/Redux/ReduxStore';
import { FeatureLevel } from '@app/Shared/Services/service.types';
import { withThemedIcon } from '@app/utils/withThemedIcon';
import { useCryostatTranslation } from '@i18n/i18nextUtil';
import { Button, ButtonVariant } from '@patternfly/react-core';
import { FileIcon, UnknownIcon, UserIcon } from '@patternfly/react-icons';
import { nanoid } from '@reduxjs/toolkit';
import { TFunction } from 'i18next';
import * as React from 'react';
import { useDispatch } from 'react-redux';
import { AutomatedAnalysisCardDescriptor } from './AutomatedAnalysis/AutomatedAnalysisCard';
import { JFRMetricsChartCardDescriptor } from './Charts/jfr/JFRMetricsChartCard';
import { MBeanMetricsChartCardDescriptor } from './Charts/mbean/MBeanMetricsChartCard';
import { DiagnosticsCardDescriptor } from './Diagnostics/DiagnosticsCard';
import { JvmDetailsCardDescriptor } from './JvmDetails/JvmDetailsCard';
import {
  SerialLayoutTemplate,
  SerialCardConfig,
  LayoutTemplateVersion,
  LayoutTemplateController,
  LayoutTemplateVendor,
  DashboardLayout,
  LayoutTemplate,
  LayoutTemplateRecord,
  CardConfig,
  CardValidationResult,
  ValidationError,
  DashboardCardDescriptor,
} from './types';

export const DashboardLayoutNamePattern = /^[a-zA-Z0-9_.-]+( [a-zA-Z0-9_.-]+)*$/;
export const LayoutTemplateDescriptionPattern = /^[a-zA-Z0-9\s.,\-'";?!@#$%^&*()[\]_+=:{}]*$/;

export const mockSerialLayoutTemplate: SerialLayoutTemplate = {
  name: 'Default',
  description: 'Default.',
  cards: [] as SerialCardConfig[],
  version: LayoutTemplateVersion['v2.3'],
};

// use a provider
export const LayoutTemplateContext = React.createContext<LayoutTemplateController>({
  selectedTemplate: undefined,
  setSelectedTemplate: () => undefined,
  isUploadModalOpen: false,
  setIsUploadModalOpen: () => undefined,
});

export const iconify = (vendor: LayoutTemplateVendor): React.ReactNode => {
  switch (vendor) {
    case LayoutTemplateVendor.CRYOSTAT:
      return <ThemedCryostatLogo />;
    case LayoutTemplateVendor.BLANK:
      return <FileIcon style={{ paddingRight: '0.3rem' }} />;
    case LayoutTemplateVendor.USER:
      return <UserIcon />;
    default:
      return <UnknownIcon />;
  }
};

export const ThemedCryostatLogo = withThemedIcon(cryostatLogo, cryostatLogoDark, 'Cryostat Logo');

export const templatize = (layout: DashboardLayout, name: string, desc?: string): LayoutTemplate => {
  return {
    name: name,
    description: desc || `Custom layout template.`,
    cards: layout.cards.map((card) => {
      const { id: _id, ...cardWithoutId } = card;
      return { ...cardWithoutId };
    }) as SerialCardConfig[],
    vendor: LayoutTemplateVendor.USER,
    version: LayoutTemplateVersion['v2.3'],
  } as LayoutTemplate;
};

export const layoutize = (template: LayoutTemplate, name: string): DashboardLayout => {
  return {
    name: name,
    cards: template.cards.map((card) => {
      return {
        ...card,
        id: `${card.name}-${nanoid()}`,
      };
    }),
    favorite: false,
  } as DashboardLayout;
};

export const recordToLayoutTemplate = (
  record: LayoutTemplateRecord,
  allTemplates: LayoutTemplate[],
): LayoutTemplate | undefined => {
  return allTemplates.find((template) => template.name === record.name && template.vendor === record.vendor);
};

export const getUniqueIncrementingName = (init = 'Custom', names: string[]): string => {
  let name: string;
  let i = 1;
  do {
    name = `${init}${i}`;
    i++;
  } while (names.includes(name));
  return name;
};

export const hasCardDescriptorByName = (name: string): boolean => {
  for (const choice of getDashboardCards()) {
    if (choice.component.cardComponentName === name) {
      return true;
    }
  }
  return false;
};

export const getCardDescriptorByName = (name: string): DashboardCardDescriptor => {
  for (const choice of getDashboardCards()) {
    if (choice.component.cardComponentName === name) {
      return choice;
    }
  }
  throw new Error(`Unknown card type selection: ${name}`);
};

export const hasCardDescriptorByTitle = (title: string, t: TFunction): boolean => {
  for (const choice of getDashboardCards()) {
    if (t(choice.title) === title) {
      return true;
    }
  }
  return false;
};

export const getCardDescriptorByTitle = (title: string, t: TFunction): DashboardCardDescriptor => {
  for (const choice of getDashboardCards()) {
    if (t(choice.title) === title) {
      return choice;
    }
  }
  throw new Error(`Unknown card type selection: ${title}`);
};

export const getDashboardCards: (featureLevel?: FeatureLevel) => DashboardCardDescriptor[] = (
  featureLevel = FeatureLevel.DEVELOPMENT,
) => {
  const cards = [
    JvmDetailsCardDescriptor,
    AutomatedAnalysisCardDescriptor,
    JFRMetricsChartCardDescriptor,
    MBeanMetricsChartCardDescriptor,
    DiagnosticsCardDescriptor,
  ];
  return cards.filter((card) => card.featureLevel >= featureLevel);
};

export const RemoveCardAction: React.FC<{ cardIndex: number }> = ({ cardIndex }) => {
  const dispatch = useDispatch();
  const { t } = useCryostatTranslation();

  const handleClick = React.useCallback(() => {
    dispatch(dashboardConfigDeleteCardIntent(cardIndex));
  }, [dispatch, cardIndex]);

  return (
    <Button onClick={handleClick} variant={ButtonVariant.danger}>
      {t('REMOVE')}
    </Button>
  );
};

export const validateCardConfig = ({ name, props }: CardConfig, cardIndex: number): CardValidationResult => {
  // Unsupported card type
  if (!hasCardDescriptorByName(name)) {
    return {
      errors: [
        {
          message: (
            <>
              Unknown card type: <code>{name}</code>.
            </>
          ),
        },
      ],
      callForAction: <RemoveCardAction cardIndex={cardIndex} />,
    };
  }

  const errs: ValidationError[] = [];

  const { propControls } = getCardDescriptorByName(name);
  const configPropKeys = Object.entries(props);

  // Missing props
  propControls
    .map((ctrl) => ctrl.key)
    .forEach((propKey) => {
      const matched = configPropKeys.find(([key, _]) => key === propKey);
      if (!matched) {
        errs.push({
          message: (
            <>
              Missing card property: <code>{propKey}</code>.
            </>
          ),
        });
      }
    });

  // Invalid prop valies
  configPropKeys.forEach(([propKey, propValue]) => {
    const matched = propControls.find((crtl) => crtl.key === propKey);
    if (matched) {
      const { values, extras, kind } = matched;
      // FIXME: Check dynamic values
      let err: ValidationError | undefined;
      if (propValue === undefined || propValue === null) {
        err = {
          message: (
            <>
              Undefined value for card property: <code>{propKey}</code>
            </>
          ),
        };
      } else if (values && Array.isArray(values) && !values.some((v) => v === propValue)) {
        err = {
          message: (
            <>
              Invalid value <code>{propValue}</code> for card property: <code>{propKey}</code>.
            </>
          ),
        };
      } else if (extras) {
        if (kind === 'number') {
          const valAsNum = Number(propValue);
          if (isNaN(valAsNum)) {
            err = {
              message: (
                <>
                  Numeric value expected but <code>{typeof propValue}</code> was given for: <>{propKey}</>.
                </>
              ),
            };
          } else {
            if (
              (extras.min !== undefined && valAsNum < extras.min) ||
              (extras.max !== undefined && valAsNum > extras.max)
            ) {
              err = {
                message: (
                  <>
                    Value exceeding limit for card property: <code>{propKey}</code>.
                  </>
                ),
              };
            }
          }
        }
      }
      err && errs.push(err);
    }
    // Unknown props are ignored
  });

  return {
    errors: errs,
    callForAction: errs.length ? <RemoveCardAction cardIndex={cardIndex} /> : undefined,
  };
};
