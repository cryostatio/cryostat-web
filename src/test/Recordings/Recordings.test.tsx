/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Recordings } from '@app/Recordings/Recordings';
import { Target } from '@app/Shared/Services/api.types';
import { defaultServices } from '@app/Shared/Services/Services';
import { cleanup, screen, within } from '@testing-library/react';
import '@testing-library/jest-dom';
import { of } from 'rxjs';
import { render, renderSnapshot } from '../utils';

jest.mock('@app/Recordings/ActiveRecordingsTable', () => {
  return {
    ActiveRecordingsTable: jest.fn((_) => {
      return <div>Active Recordings Table</div>;
    }),
  };
});

jest.mock('@app/Recordings/ArchivedRecordingsTable', () => {
  return {
    ArchivedRecordingsTable: jest.fn((_) => {
      return <div>Archived Recordings Table</div>;
    }),
  };
});

jest.mock('@app/TargetView/TargetView', () => {
  return {
    TargetView: jest.fn((props) => {
      return (
        <div>
          {props.pageTitle}
          {props.children}
        </div>
      );
    }),
  };
});

const mockFooTarget: Target = {
  connectUrl: 'service:jmx:rmi://someFooUrl',
  alias: 'fooTarget',
  labels: [],
  annotations: {
    cryostat: [],
    platform: [],
  },
};

jest.spyOn(defaultServices.target, 'target').mockReturnValue(of(mockFooTarget));

jest
  .spyOn(defaultServices.api, 'isArchiveEnabled')
  .mockReturnValueOnce(of(true)) // has the correct title in the TargetView
  .mockReturnValueOnce(of(true)) // handles the case where archiving is enabled
  .mockReturnValueOnce(of(false)) // handles the case where archiving is disabled
  .mockReturnValue(of(true)); // others

describe('<Recordings />', () => {
  afterEach(cleanup);

  it('has the correct title in the TargetView', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <Recordings />,
          },
        ],
      },
    });

    expect(screen.getByText('Recordings')).toBeInTheDocument();
  });

  it('handles the case where archiving is enabled', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <Recordings />,
          },
        ],
      },
    });

    expect(screen.getByText('Active Recordings')).toBeInTheDocument();
    expect(screen.getByText('Archived Recordings')).toBeInTheDocument();
  });

  it('handles the case where archiving is disabled', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <Recordings />,
          },
        ],
      },
    });

    expect(screen.getByText('Active Recordings')).toBeInTheDocument();
    expect(screen.queryByText('Archived Recordings')).not.toBeInTheDocument();
  });

  // useNavigate() is mocked. Can't switch tab
  it.skip('handles updating the activeTab state', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <Recordings />,
          },
        ],
      },
    });

    // Assert that the active recordings tab is currently selected (default behaviour)
    let tabsList = screen.getAllByRole('tab');

    let firstTab = tabsList[0];
    expect(firstTab).toHaveAttribute('aria-selected', 'true');
    expect(within(firstTab).getByText('Active Recordings')).toBeTruthy();

    let secondTab = tabsList[1];
    expect(secondTab).toHaveAttribute('aria-selected', 'false');
    expect(within(secondTab).getByText('Archived Recordings')).toBeTruthy();

    // Click the archived recordings tab
    await user.click(screen.getByText('Archived Recordings'));

    // Assert that the archived recordings tab is now selected
    tabsList = screen.getAllByRole('tab');

    firstTab = tabsList[0];
    expect(firstTab).toHaveAttribute('aria-selected', 'false');
    expect(within(firstTab).getByText('Active Recordings')).toBeTruthy();

    secondTab = tabsList[1];
    expect(secondTab).toHaveAttribute('aria-selected', 'true');
    expect(within(secondTab).getByText('Archived Recordings')).toBeTruthy();
  });

  it('renders correctly', async () => {
    const tree = await renderSnapshot({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <Recordings />,
          },
        ],
      },
    });
    expect(tree?.toJSON()).toMatchSnapshot();
  });
});
