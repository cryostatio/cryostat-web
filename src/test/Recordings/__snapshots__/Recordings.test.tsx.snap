// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recordings /> renders correctly 1`] = `
<div>
  Recordings
  <article
    className="pf-c-card"
    data-ouia-component-id="OUIA-Generated-Card-5"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe={true}
    id=""
  >
    <div
      className="pf-c-card__body"
    >
      <div
        className="pf-c-tabs"
        data-ouia-component-id="OUIA-Generated-Tabs-4"
        data-ouia-component-type="PF4/Tabs"
        data-ouia-safe={true}
        id="recordings"
        onSelect={[Function]}
      >
        <ul
          className="pf-c-tabs__list"
          onScroll={[Function]}
          role="tablist"
        >
          <li
            className="pf-c-tabs__item pf-m-current"
            role="presentation"
          >
            <button
              aria-controls="pf-tab-section-0-active-recordings"
              aria-selected={true}
              className="pf-c-tabs__link"
              data-ouia-component-type="PF4/TabButton"
              data-ouia-safe={true}
              id="pf-tab-0-active-recordings"
              onClick={[Function]}
              role="tab"
              type="button"
            >
              <span
                className="pf-c-tabs__item-text"
              >
                Active Recordings
              </span>
            </button>
          </li>
          <li
            className="pf-c-tabs__item"
            role="presentation"
          >
            <button
              aria-controls="pf-tab-section-1-archived-recordings"
              aria-selected={false}
              className="pf-c-tabs__link"
              data-ouia-component-type="PF4/TabButton"
              data-ouia-safe={true}
              id="pf-tab-1-archived-recordings"
              onClick={[Function]}
              role="tab"
              type="button"
            >
              <span
                className="pf-c-tabs__item-text"
              >
                Archived Recordings
              </span>
            </button>
          </li>
        </ul>
      </div>
      <section
        aria-labelledby="pf-tab-0-active-recordings"
        className="pf-c-tab-content"
        data-ouia-component-type="PF4/TabContent"
        data-ouia-safe={true}
        hidden={false}
        id="pf-tab-section-0-active-recordings"
        role="tabpanel"
        tabIndex={0}
      >
        <div>
          Active Recordings Table
        </div>
      </section>
      <section
        aria-labelledby="pf-tab-1-archived-recordings"
        className="pf-c-tab-content"
        data-ouia-component-type="PF4/TabContent"
        data-ouia-safe={true}
        hidden={true}
        id="pf-tab-section-1-archived-recordings"
        role="tabpanel"
        tabIndex={0}
      >
        <div>
          Archived Recordings Table
        </div>
      </section>
    </div>
  </article>
</div>
`;
